flink-streaming-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducerBase.java

```diff
@@ -276,6 +315,38 @@ public void close() throws Exception {
 		checkErroneous();
 	}
 
+	// ------------------- Logic for handling checkpoint flushing -------------------------- //
+
+	private void acknowledgeMessage() {
+		if(!flushOnCheckpoint) {
+			// the logic is disabled
+			return;
+		}
+		pendingRecords--;
```

Comment:
I think it is necessary to guard against concurrent changes from the `invoke` method.